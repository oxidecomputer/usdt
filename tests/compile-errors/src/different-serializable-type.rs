//! Test that passing a type that is serializable, but not the same concrete type as the probe
//! signature, fails compilation.

// Copyright 2021 Oxide Computer Company

#![feature(asm)]
#![cfg_attr(target_os = "macos", feature(asm_sym))]

#[derive(serde::Serialize)]
struct Expected {
    x: u8
}

#[derive(serde::Serialize)]
struct Different {
    x: u8
}

#[usdt::provider]
mod my_provider {
    use crate::Expected;
    fn my_probe(_: Expected) {}
}

fn main() {
    my_provider::my_probe!(|| Different { x: 0 });
}
