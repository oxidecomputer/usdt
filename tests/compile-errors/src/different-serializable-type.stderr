error[E0277]: the trait bound `Different: Borrow<Expected>` is not satisfied
  --> src/different-serializable-type.rs:28:1
   |
28 | #[usdt::provider]
   | ^^^^^^^^^^^^^^^^^ the trait `Borrow<Expected>` is not implemented for `Different`
...
35 |     my_provider::my_probe!(|| Different { x: 0 });
   |     --------------------------------------------- in this macro invocation
   |
note: required by a bound in `__usdt_private_my_provider_my_probe_type_check`
  --> src/different-serializable-type.rs:28:1
   |
28 | #[usdt::provider]
   | ^^^^^^^^^^^^^^^^^ required by this bound in `__usdt_private_my_provider_my_probe_type_check`
...
35 |     my_provider::my_probe!(|| Different { x: 0 });
   |     --------------------------------------------- in this macro invocation
   = note: this error originates in the macro `my_provider::my_probe` (in Nightly builds, run with -Z macro-backtrace for more info)
